### PR TITLE
clear pending atts after syncing block

### LINKED ISSUE #9662: Process attestations after processing block
# 🐞 Bug Report

### The issue
Prysm does not process early attestations in the current slot, but rather in the next slot. 

### Description of the problem
Consider the following timeline

1. Beacon node receives an attestation `A` for the current block, 1 second into the slot. 
2. Beacon node does not have the current block, so the attestation is put into a queue. 
3. Beadon node receives the block 3 seconds into the slot and processes it by 3.2 seconds. 
4. Beacon node receives an attestation `B` for the current block at 4.2 seconds and since it has the block it processes it immediately. 

Then the late attestation `B` is processed and gossiped, but the attestation `A` is not processed until the queue is cleared at the beginning of the next slot. 

### Detailed description of the problem:
Blocks in prysm are handled by [onBlock](https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/blockchain/process_block.go#L86), this function processes the block and the attestations within it, updates the head, the justification and finalized points, and performs fork-choice operations related to the new block we just synced. But it **does not call to clear attestation queues for this received slot**. 

The way attestations are processed is by running a [routine](https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/blockchain/receive_attestation.go#L101) that has the following loop in it

```go
       st := slots.NewSlotTicker(s.genesisTime, params.BeaconConfig().SecondsPerSlot)
	for {
		select {
		case <-s.ctx.Done():
			return
		case <-st.C():
			// Continue when there's no fork choice attestation, there's nothing to process and update head.
			// This covers the condition when the node is still initial syncing to the head of the chain.
			if s.cfg.AttPool.ForkchoiceAttestationCount() == 0 {
				continue
			}
			s.processAttestations(s.ctx)
			if err := s.updateHead(s.ctx, s.getJustifiedBalances()); err != nil {
				log.Warnf("Resolving fork due to new attestation: %v", err)
			}
		}
	}
```
The key point is that 1) this happens once per slot because of the `NewSlotTicker` and 2) `processAttestations` is what handles the delay in processing them. 

### Possible solutions

I can see two options here, the optimal solution from the perspective of the network, if feasible, would be to call `processAttestations` right after updating the head, for example it can be done [here](https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/blockchain/receive_block.go#L54). 

Option 2 would be to at least make sure that we process these attestations sometime between 4 seconds and 7.5 seconds so as to at least give a minimal chance that these attestations will be seen by the aggregators of the current slot instead of the next one. 

Option 1 needs to be weighted against performance impact. I favor this option because of the following:
 
- Right after processing a block and updating head, not a busy time for the beacon, in fact this problem happens more often in slot 0 and after receiving the block for slot 0 we have already finished epoch transition and are idle essentially. 
- It makes sure that it processes early attestations immediately and therefore gossips them and can alleviate the problem of slot 0 blocks being late. 
- In a highly forked situation, we may update the head wildly, but we will not be calling this extra processing more than once per slot anyway, so it does not provide us with a problem when we are highly forked. In the worst possible case of a context deadline, we will automatically revert to the previous behavior since the queue will be cleared in the next tick. 

Option 2 is simplistic but it provides minimum relief: we can just replace the tick there by `params.BeaconConfig().SecondsPerSlot / 2 ` and at least we would be gossiping 2 seconds before aggregation which gives some chances for those attestations. I tried on pyrmont attesting at that time and wasn't penalized.

### PR DESCRIPTION
**Pull request by potuz**
Call `processPendingAtts` right after receiving and processing a block. this calls only once per block received instead of twice per slot. There is no point in calling this routine at a fixed time now since we won't be able to process any attestations if we hadn't synced the block. 

This PR separates the two functions of the pending attestation queue, namely requesting pending blocks, and processing attestations, to be made in parallel and on demand when validating the newly arrived attestation and right after syncing a block respectively. It follows the design in https://hackmd.io/@potuz/B1yC3gFNY

Credit goes to @nisdas that traced the problem to `attest-timely` used in conjunction with the pending attestation queue design of prysm. 

Fixes #9662

### CODE REVIEW

File: beacon-chain/sync/pending_attestations_queue.go [hunk 1]
```diff
@@ -18,161 +17,144 @@ import (
...
+		log.WithFields(logrus.Fields{
+			"currentSlot": s.cfg.Chain.CurrentSlot(),
+			"attSlot":     att.Message.Aggregate.Data.Slot,
+			"blockRoot":   hex.EncodeToString(bytesutil.Trunc(root[:])),
+		}).Debug("Requesting block for pending attestation")
+
+		return s.sendBatchRootRequest(context.Background(), [][32]byte{root}, randGen)
 	}
 
+	defer s.pendingAttsLock.Unlock()
```
Comment (nisdas) [inline beacon-chain/sync/pending_attestations_queue.go:147]: This will be a deadlock if we exit early above when sending the batch request.

Comment (potuz) [reply beacon-chain/sync/pending_attestations_queue.go:147]: don't I unlock explicitly in that case?

**Comment by rauljordan** [comment]
Is this still relevant @terencechain @potuz ?

**Comment by potuz** [comment]
> Is this still relevant @terencechain @potuz ?

Yes, it's important to prevent missed attestations on slot zero for attesters running with `--attest-timely`. It's being tested in Prater already for a few days. We'll test it longer since this is not gated through a flag.

**Review by terencechain — APPROVED** [review]

**Comment by rauljordan** [comment]
Hi @potuz just checking in on the status of this

**Comment by potuz** [comment]
> Hi @potuz just checking in on the status of this

We tested in Prater, didn't find any sign that this PR breaks anything, but got slightly worse attestation performance on our validators with this in combination with --attest-timely. We didn't do extensive testings. Decided to put it on the next major upgrade cc @nisdas

**Comment by potuz** [comment]
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed in 7 days if no further activity occurs. Thank you for your contributions.

mind your own business :)

**Comment by potuz** [comment]
Close as @nisdas sat on this for long enough

**Pull request closed without merging**